src/token_contract/src/main.nr

```diff
@@ -676,4 +697,98 @@ pub contract Token {
+        let max_shares = _max_mint();
+        assert(assets <= max_shares, "Max mint exceeded");
+
+        let total_assets = Token::at(storage.asset.read()).balance_of_public(context.this_address()).call(&mut context);
+
+        let ten_to_the_power_of_decimal_offset = 10.pow_32(0);  
+        let _shares = assets * (storage.total_supply.read() + ten_to_the_power_of_decimal_offset) / (total_assets + 1)
+
+        // Take the asset from the caller
+        Token::at(storage.asset.read()).transfer_public_to_public(context.msg_sender(), context.this_address(), _shares, nonce).call(&mut context);
+
+        // mint shares to the receiver
+        _increase_public_balance(storage.public_balance, receiver, _shares);
+        _increase_total_supply(storage.total_supply, _shares);
+    }
+
+    #[private]
+    fn deposit_private_to_public(assets: u128, receiver: AztecAddress, total_supply: u128, total_assets: u128, nonce: Field) {
+        let max_shares = _max_mint();
+        assert(assets <= max_shares, "Max mint exceeded");
+
+        let ten_to_the_power_of_decimal_offset = 10.pow_32(0);  
+        let _shares = assets * (total_supply + ten_to_the_power_of_decimal_offset) / (total_assets + 1)
+
+        // Take the asset from the caller
+        Token::at(storage.asset.read()).transfer_private_to_public(context.msg_sender(), context.this_address(), assets, nonce).call(&mut context);
+        
+        Token::at(context.this_address()).validate_total_supply_and_assets_and_increase_internal(total_supply, total_assets, _shares).enqueue(
+            &mut context,
+        );
+
+        // mint shares to the receiver in public
+        Token::at(context.this_address()).increase_public_balance_internal(receiver, _shares).enqueue(
+            &mut context,
+        );
+    }
+
+    #[private]
+    fn deposit_private_to_private(assets: u128, receiver: AztecAddress, total_supply: u128, total_assets: u128, nonce: Field) {
+        let max_shares = _max_mint();
+        assert(assets <= max_shares, "Max mint exceeded");
+
+        let ten_to_the_power_of_decimal_offset = 10.pow_32(0);  
+        let _shares = assets * (total_supply + ten_to_the_power_of_decimal_offset) / (total_assets + 1)
+
+        // Take the asset from the caller
+        Token::at(storage.asset.read()).transfer_private_to_public(context.msg_sender(), context.this_address(), assets, nonce).call(&mut context);
+
+        Token::at(context.this_address()).validate_total_supply_and_assets_and_increase_internal(total_supply, total_assets, _shares).enqueue(
+            &mut context,
+        );
+
+        // mint shares to the receiver
+        _increase_private_balance(
+            &mut context,
+            storage.private_balance,
+            receiver,
+            _shares,
+        );
+    }
+
+    #[public]
+    fn validate_total_supply_and_assets_and_increase_internal(total_supply_received: u128, total_assets_received: u128, amount: u128) {
```

Comment:
hmm interesting that we're merging "validate" + "mint", the bad part is that we're increasing the private balance in private so we don't actually know the result of this calc, we should be validating the ratio, and optimistically minting, i agree with that